### DISCUSSION TITLE
Ideas for undergraduate project

### DISCUSSION THREAD

**Discussion by erelsgl** [cvxpy/cvxpy#1890] [General]
I have three undergraduate 3rd-year students for computer science. As their final year programming project, they would like to contribute to `cvxpy`. The contribution should be substantial enough for an annual project (e.g. not just documentation or fixing minor bugs), but it should not be too mathematically heavy - these are students for computer science, not mathematics. What are some issues that could benefit from such a contribution?

**Comment by SteveDiamond** [comment]
The first thing that comes to mind is adding more functionality for visualizing or pretty printing cvxpy problems. It's quite difficult right now to analyze a problem object without a script to construct it IMO, and better visualization/printing would make a big difference. @phschiele had some visualization functionality he showed me that might be relevant.

**Reply by erelsgl** [reply]
This sounds great. I will be happy to see some examples of what such visualizations could look like.

**Reply by SteveDiamond** [reply]
Excellent! @phschiele could you share that visualization of linop trees you showed me?

**Reply by phschiele** [reply]
@erelsgl That sounds like an awesome initiative! I had written a short snippet that plotted the canonicalization tree, which looked something like this:
![image](https://user-images.githubusercontent.com/44360364/190223661-6197eeed-4d3c-4ce3-8161-72ba3fdf18b4.png)
It would be more useful to apply such recursive plotting to cvxpy expressions, resulting in a tree similar to https://dcp.stanford.edu/analyzer (ideally showing curvature, sign, and monotonicity properties of each node). As a first step, a static plot like the above could be generated, but I'm sure it would not be too hard to make it interactive, e.g. using `dash`. 
Either way, if this ends up as the project you are deciding to do, I'd be happy to join for a kick-off/brainstorming session.

**Reply by erelsgl** [reply]
@phschiele 
I will be happy to set a Zoom meeting for brainstorming. When are you available?

**Reply by erelsgl** [reply]
@phschiele also, regarding your example tree, what code exactly would produce it, in your view?

**Reply by SteveDiamond** [reply]
I can potentially join the zoom meeting as well. Are you on the discord (https://discord.gg/4urRQeGBCr)? It will be easier to coordinate there.

**Reply by erelsgl** [reply]
I could not connect to discord. But maybe we can coordinate with Doodle: https://doodle.com/meeting/participate/id/egZJnmGd

**Reply by SteveDiamond** [reply]
Ok that's fine. These times are not great for PST, where Philipp and I are. Is this the only time range that works for you?

**Comment by erelsgl** [comment]
Ah, sorry, I did not know you are in PST. I changed the times.

**Reply by erelsgl** [reply]
OK! Monday 31/10 at 9:00 AM PDT looks like the best option. Here is the Zoom link: https://ariel-ac-il.zoom.us/j/3628593340

**Reply by SteveDiamond** [reply]
You mean 9am PDT right? That's the earliest time monday in the doodle.

**Reply by erelsgl** [reply]
Yes, 9am PDT. Fixed

**Reply by phschiele** [reply]
Great, looking forward!

**Reply by erelsgl** [reply]
Actually, I have to move it earlier - to **Sunday** same time (9am PDT).  I hope this time is OK too.

**Reply by SteveDiamond** [reply]
That's fine.

**Comment by phschiele** [comment]
Here is the code for plotting the lin op tree:

```py
import pydot

from cvxpy.lin_ops.lin_utils import get_id


def plot_tree(op_list):
    for i, tree in enumerate(op_list):
        mapping = get_all_args(tree)
        if isinstance(mapping, tuple):
            pass
        else:
            graph = pydot.Dot(graph_type='graph')
            visit(mapping, graph)
            graph.write_png(f'{i}.png')


def get_all_args(expression):
    if isinstance(expression, tuple):
        return get_id(), str(expression)
    elif expression.args == [] or expression.data == 0:
        if expression.type == "variable":
            return get_id(), expression.type + f" {expression.data} " + str(expression.shape)
        elif expression.type == "param":
            return get_id(), expression.type + f" {expression.data} " + str(expression.shape)
        else:
            return get_id(), expression.type + " " + str(expression.shape)
    else:
        return {(get_id(), expression.type): [get_all_args(x) for x in expression.args + [expression.data] if
                                              x is not None]}


def draw(parent_name, child_name, graph):
    if not parent_name[0] in graph.get_nodes():
        graph.add_node(pydot.Node(str(parent_name[0]), label=parent_name[1]))
    if not child_name[0] in graph.get_nodes():
        graph.add_node(pydot.Node(str(child_name[0]), label=child_name[1]))
    edge = pydot.Edge(parent_name[0], child_name[0])
    graph.add_edge(edge)


def visit(node, graph, parent=None):
    for k, v in node.items():
        if parent is not None:
            draw(parent, k, graph)
        for j in v:
            if isinstance(j, dict):
                visit(j, graph, parent=k)
            else:
                draw(k, j, graph)

```
